elastic/detection-rules
# [FR] Add `source_updated_at` to Rule Schema as a Build Time Field

## Issues
* https://github.com/elastic/detection-rules/issues/2826

## Summary
This pull request adds `source_updated_at` to the `BaseRuleSchema`, defines a minimum compatible stack for this field and dynamically generates it on rule loading. For more details, please refer to the issue referenced.

## Testing

### Testing Rule Version Adjustments
Based on the suggested changes, rule versions should not change even though `source_updated_at` is added to the final rule object. This is because we remove this key:value pair from the dictionary that is SHA256 hashed. This addresses the versioning concern of two separate versions between branches as it is not taken into consideration. We did this for a few reasons:

* We understand it is an exception, but the date is pulled from metadata, we never expect a rule author to add this in the rule schema, and it is controlled solely by us with no influence upstream
* We need a mechanism to bypass version breaking changes if we can and this supplies one; we also do this in endpoint rules as well so the approach is not net new to our repositories and logic

1. Open `rule.py` and comment out L982 (`self._convert_add_source_updated_at(obj)`)
2. Run `python -m detection_rules view-rule /Users/tdejesus/code/src/detection-rules/rules/windows/persistence_startup_folder_scripts.toml`
3. Note the version is `109`
4. Uncomment step 1. and run the CLI command again
5. Note the version is still `109`

This is because regardless if the field is added or note, we simply remove it if found all the time when calculating SHA256. This would backport and happen for each branch, thus version would be the same but the field would be included in the package moving forward.

For compatibility, the field is still added to the `BaseRuleData` dataclass to support importing the rule.

### Testing Field is Dynamically Generated
Remember that this field will only generate if the stack version (reflected by `packages.yml`) is `>=8.12`. Therefore we need to test that the field is included AND not included on on older branches

CLI Command: `python -m detection_rules view-rule /Users/tdejesus/code/src/detection-rules/rules/windows/persistence_startup_folder_scripts.toml`

The `source_updated_at` field should be visible in the output.

To re-create this for a lower stack version to ensure it does not appear, copy the `rule.py` and `definitions.py` files to a separate folder, checkout an older branch, move them back and then run the CLI command again; `source_updated_at` should not be added to the final rule output. 

This is because of the logic [here](https://github.com/elastic/detection-rules/blob/99059e26cf97d743bc75a4563ca630d8cc8c4e22/detection_rules/rule.py#L1000). This is also why the `TestBuildTimeFields` unit test class was removed as it is redundant and we do this check for every build time field.
